d3a_api_client/grid_fee_calculation.py

```diff
@@ -0,0 +1,102 @@
+import logging
+from copy import copy
+
+from d3a_interface.utils import key_in_dict_and_not_none
+
+
+class GridFeeCalculation:
+
+    def __init__(self):
+        self.latest_grid_stats_tree = {}
+        self.paths_to_root_mapping = {}
+        self.market_name_grid_fee_mapping = {"last_market_fee": {},
+                                             "next_market_fee": {}}
+
+    def _handle_grid_stats(self, message):
+        for device_event in message["content"]:
+            if device_event["event"] == "grid stats":
```

Comment:
```suggestion
            if device_event["event"] == "grid_stats":
```
Minor , we can skip it .

Reply:
No, you are right! Its better to have an underscore here.

Reply:
In addition to that, is there a need for an extra event, other than the market_cycle one ? If the problem is when to decide to calculate the grid fees or not, it can be done fairly quickly by checking if the grid_stats_tree contains children or not, removing the need for another event type.

Reply:
If you have a look into the d3a PR ( https://github.com/gridsingularity/d3a/pull/1045 )
I didn't add a new event type, I only marked the grid_stats by setting the "event" member in the market_event.
The one grid_stats information is prepended to all market_events of one aggregator.
So if the aggregator controls 2 devices, it will get a list of two market_cycle_events plus one grid_stats information.
Now I only explicitly market it as such. Let me know if you need more info.